src/uu/ls/src/ls.rs

```diff
@@ -1038,12 +1039,43 @@ pub fn uumain(args: impl uucore::Args) -> i32 {
     list(locs, Config::from(matches))
 }
 
+/// Represents a Path along with it's associated data
+/// Any data that will be reused several times makes sense to be added to this structure
+/// Caching data here helps eliminate redundant syscalls to fetch same information
+struct PathData {
+    // Result<MetaData> got from symlink_metadata() or metadata() based on config
+    md: std::io::Result<Metadata>,
```

Comment:
Would it make sense to return this `Result` on `PathData::new`? The `fs::metadata` call fails if the path does not exist or the user does not have permissions to read the metadata, in which case I think we want to log that and then not consider the file for any later calls. So the new method would look rougly like this:
```rust
impl PathData {
    fn new(p_buf: PathBuf, config: &Config, command_line: bool) -> std::io::Result<Self> {
        let md = get_metadata(&p_buf, config, command_line)?;
        ...
        Ok(Self{...})
    }
}
```
And then we don't have to check for errors from the metadata later on.

Reply:
Yes, that's another good idea.
Need to check where / when the error is expected to be logged, when we print the details of the file (after sorting) or if it is fine to log all errors upfront.

For metadata, I had another optimization in mind where if the config does not include any options that require metadata details, we could skip get_metadata call altogether (type for the md field could then become Option<MetaData>)

Reply:
Ah yeah, that's a neat idea! An easy way to implement this might be to make a method `PathData::md()` that retrieves the metadata on the first call and then stores it in the struct and reuses that value when it's called again.